`__getstate__` and `__setstate__` consolidation

s0undt3ch opened the pull request:
### What does this PR do?
The `Process` class will now take care of all `__getstate__` and `__setstate__`

This PR will then ensure that any forked process, on windows, and of course all platforms which support forking, will behave properly **without** having to implement their own get and set state functions, unless they need to extend the base functionality provided in out Process class.

### Previous Behavior
All subclasses had to implement `__getstate__` and `__setstate__`

### New Behavior
The `Process` class will now take care of all `__getstate__` and `__setstate__`

### Tests written?

No

### Commits signed with GPG?

Yes

waynew commented on salt/engines/__init__.py:
```diff
@@ -81,33 +81,6 @@ def __init__(self, opts, fun, config, funcs, runners, proxy, **kwargs):
-            'log_queue': self.log_queue,
-            'log_queue_level': self.log_queue_level
-        }
-
```
Not sure that I understand - how do all of these things end out getting called? They're quite different than what's in the other files.

s0undt3ch replied:
On windows, and windows alone, the `__new__` method will keep a copy of the `*args` and `**kwargs` used to instantiate the class.
Then, again, only on windows, the get and set state methods will use that information to re-instsntiate the class.

s0undt3ch replied:
[Here](https://github.com/saltstack/salt/pull/55793/files#diff-0eb908cc193e4bf8dd9f96567a3abfebR741)

s0undt3ch commented:
This PR will then ensure that any forked process, on windows, and of course all platforms which support forking, will behave properly **without** having to implement their own get and set state functions.

s0undt3ch commented:
Unless they need to extend the base functionality provided in out Process class.

s0undt3ch commented:
I need to make sure any extra methods for post fork and pre shutdown are also carried over to the new process.

s0undt3ch committed:
Allow registering after fork and finalize methods

s0undt3ch committed:
No claim was made so the file does not exist

s0undt3ch committed:
A path is expected, give it a proper path.

s0undt3ch commented:
re-run full all

dwoz reviewed (commented):
I'm not a fan of this. It seems reasonable to me to expect sub-classes of Process to implement their own `__getstate__` and `__setstate__` rather than trying to be smart about it in the constructor.

s0undt3ch commented:
As we discussed, it gets us, at least, logging on those processes for free.

s0undt3ch committed:
The `Process` class will now take care of all `__getstate__` and `__setstate__`

s0undt3ch committed:
Add `salt.utils.plaform.spawnin_platform()`

This allows checking for platforms which default to `spawn`
multiprocessing processes instead of doing a check for each platform
that defaults to `spawn` depending on the python version.

s0undt3ch commented:
Conflicts resolved

s0undt3ch commented:
The pre-commit issue is being solved in https://github.com/saltstack/salt/pull/60254

s0undt3ch commented:
re-run full all

s0undt3ch commented:
re-run full all

s0undt3ch commented:
re-run full all

s0undt3ch commented:
re-run full all

s0undt3ch commented:
re-run full all

s0undt3ch commented:
re-run full all

s0undt3ch commented:
re-run full all

s0undt3ch commented:
re-run full all

s0undt3ch commented:
re-run full all

s0undt3ch commented:
re-run full all

waynew commented on salt/utils/platform.py:
```diff
@@ -209,3 +210,12 @@ def is_aarch64():
     return platform.machine().startswith("aarch64")
+
+
+def spawning_platform():
```
**question** is it better to gate this behind a function or do something like this at the module level?

```
SPAWNING_PLATFORM = multiprocessing.get_start_method(allow_none=False) == "spawn"
```

I don't know if it's called enough to make a difference :shrug:

s0undt3ch replied:
No, the method can be changed programmatically at run time. A global variable would only pick the initial config value.

s0undt3ch replied:
@waynew are we good to resolve this review comment?

s0undt3ch replied:
Also, if you notice, this is probably the only function which doesnt have the cached decorator, that was the reason behind it.
